feat: add model streaming

gt2345 opened the pull request:
## Description

<!---
Lead with the intended commit body in this description field. For breaking
changes, please include "BREAKING CHANGE:" at the beginning of your commit
body.  At a minimum, this section should include a bracketed reference to the
Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly
into the description field.
-->
Add model streaming, this is not yet exposed to the WebUI.
There will be a separate PR for EE to handle rbac.


## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot
as appropriate.  Reviewers may ask questions about this test plan to ensure
adequate manual coverage of changes.
-->
Connect to `ws://localhost:8080/stream`, should be able to subscribe and unsubscribe for `models`.



## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for
particularly tricky bits of code that could use extra scrutiny, historical
context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->
There is a separate ticket for adding unit test, so unit test is not yet included in this PR. 



## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
<!---
Retain the relevant line and replace 000 with ticket number.

DET-000
MLG-000
WEB-000
DESIGN-000
No Ticket
--->
[MD-261](https://hpe-aiatscale.atlassian.net/browse/MD-261)

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

-->


[MD-261]: https://hpe-aiatscale.atlassian.net/browse/MD-261?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

corban-beaird commented on master/static/migrations/20240304145201_add-models-trigger.tx.up.sql:
```diff
@@ -0,0 +1,71 @@
+DROP TRIGGER IF EXISTS stream_model_trigger_seq ON models;
+CREATE TRIGGER stream_model_trigger_seq
+    BEFORE INSERT OR UPDATE OF
+    name, description, creation_time, last_updated_time, metadata, labels, user_id, archived, notes, workspace_id
```
it's likely outside the scope of this ticket; however, due to the limitations of postgres notify/listen queues, we'll want to only communicate over these channels using filterable columns. It doesn't look like the work on making the transition to the strategy is complete.

That being said, this will need to be adjusted with the work so that we don't potentially exceed that 8k character limit with user defined fields

gt2345 replied:
For models, I think it should be safe to exclude `creation_time` and `user_id` from the trigger, but do you mean we also want to exclude `metadata` and `notes`? I know we have a [ticket](https://hpe-aiatscale.atlassian.net/browse/MD-266) to address the character limit, I wonder if that's still valid, maybe @jgongd has more information?

jgongd replied:
The ticket is related to this issue. I plan to add unit tests next before focusing on this ticket. Let me know if you think this should take priority, and I'll be happy to tackle it first.

corban-beaird reviewed (commented):
The streaming updates code looks solid, could integration tests be added?

gt2345 commented:
> The streaming updates code looks solid, could integration tests be added?

Thanks! I refactored the `project` entity to use common utility functions, and I also added an integration test for streaming multiple entities.

azhou-determined commented:
should model streaming be added to the python streaming client `_client.py` as well? or is that not relevant to this PR?

azhou-determined reviewed (approved):
lgtm. commented a few suggestions/questions but up to you whether they're relevant.

gt2345 commented:
> a few general design questions wrt streaming updates and not in the scope of this PR (feel free to redirect me if this isn't the right place to ask these, just some thoughts i had while reading through the code as a newcomer to streaming updates 😄 ):
> 
> 1. it seems like streaming can only happen on a single table, and events returned to the client are also scoped to a single table. what's the plan for more complicated streaming use cases in the future? say for example i want to listen for the latest validation metric for a given trial, or the best checkpoint for an experiment (something our SDK/CLI/REST APIs currently support).
>    in the current streaming updates world, this means i would have to listen for updates on the experiments, trials, checkpoints, and also metrics tables? or do we intend to allow joins, either in the `SubscriptionSpec`s or the returned events? i.e. if i subscribe to the `runs` table, there's a foreign key `best_validation_id` to the metrics table, and that FK row automatically gets joined and included in the event.
> 2. is the assumption that streaming is mainly for listening for updates for an object that has already been created? i'm realizing that for a lot of these `SubscriptionSpec`s, we're gonna require IDs, but in order to know the IDs, you'd have to have created the object before you start streaming -- for some tables this means you basically can't listen for creation/insert events. in this case i wonder if broader filters in the `SubscriptionSpec` might be useful, like `created_ts > now()`.

Hi @azhou-determined , thanks for your review. To answer you questions: 
1. One subscription should only target one table in database, the join is supposed to happen at the client side. Just like you said, we can listen for updates for multiple tables, then join data at the front end. 
2. The streaming updates only serves read, for creating/updating objects, we are still using the rest apis. Could you please expand on the kinds of tables we can't listen for creation/insert events?

corban-beaird commented on master/internal/stream/models.go:
```diff
@@ -0,0 +1,233 @@
+// ModelSubscriptionSpec is what a user submits to define a Model subscription.
+//
+// determined:stream-gen source=client
+type ModelSubscriptionSpec struct {
```
There's hesitancy to allow for subscribing based on model name due to limitations on postgres notify/listen queues. To allow for subscribing based on model names, we'd have to enforce an maximum length. The current maximum size of a model name is only restricted by the index limit in postgres as of right now (8191 bytes), which will exceed the capacity of the postgres queues.

corban-beaird commented on master/internal/stream/models.go:
```diff
@@ -0,0 +1,233 @@
+// ModelSubscriptionSpec is what a user submits to define a Model subscription.
+//
+// determined:stream-gen source=client
+type ModelSubscriptionSpec struct {
```
I don't see an issue with allow subscription based on user id; assuming RBAC is implemented in EE. 

I would be hesitant to allow label-based subscriptions for the same reason as above, especially since there doesn't seem to be a limit on the length of a tag.

corban-beaird commented on master/internal/stream/models.go:
```diff
@@ -0,0 +1,260 @@
+type ModelSubscriptionSpec struct {
+	WorkspaceIDs []int    `json:"workspace_ids"`
+	ModelIDs     []int    `json:"model_ids"`
+	ModelNames   []string `json:"model_names"`
```
Are model names being restricted in length to ensure that we don't exceed the postgres payload limit?

gt2345 replied:
I believe not, thanks for your input, I decided to remove subscription by model name

corban-beaird reviewed (approved):
Looks great! :D
